### PR TITLE
Fix stmt/stmt2 should not insert disordered or duplicate data

### PR DESCRIPTION
**Pull request by haolinw**
# Description

- Enable client dedup for stmt/stmt2 inserts in interlace mode
- Not allow duplicate table names in stmt2 bind with interlace mode

# Checklist

Please check the items in the checklist if applicable.

- [ ] Is the user manual updated?
- [ ] Are the test cases passed and automated?
- [ ] Is there no significant decrease in test coverage?

### CODE REVIEW

File: source/client/src/clientMain.c [hunk 1]
```diff
@@ -2206,6 +2234,8 @@ int taos_stmt2_bind_param(TAOS_STMT2 *stmt, TAOS_STMT2_BINDV *bindv, int32_t col
     }
   }
 
+  tSimpleHashCleanup(hashTbnames);
```
Comment (Pengrongkun) [inline source/client/src/clientMain.c:2229]: 这块有内存泄漏风险，建议用goto保证即使上面报错也不会直接返回，而是跳转到结尾做一次内存释放。可以参考其他地方goto进行修改

File: source/client/src/clientMain.c [hunk 2]
```diff
@@ -2166,9 +2166,37 @@ int taos_stmt2_bind_param(TAOS_STMT2 *stmt, TAOS_STMT2_BINDV *bindv, int32_t col
...
+  /*
+   * Insert with interlace mode doesn't allow duplicate table names per bind,
+   * so initialize a tbnames hash table for duplication check. 
+   * 
+   * Note, STMT doesn't have such issue as it's unable to bind multiple tables.
+   */
+  SSHashObj *hashTbnames = tSimpleHashInit(100, taosGetDefaultHashFunction(TSDB_DATA_TYPE_VARCHAR));
+  if (NULL == hashTbnames) {
+    tscError("stmt2 bind failed: out of memory");
+    terrno = TSDB_CODE_OUT_OF_MEMORY;
```
Comment (Pengrongkun) [inline source/client/src/clientMain.c]: 这块好像潘魏讲过不要这么写：https://taosdata.feishu.cn/wiki/MvCGwdEqwieDcwkQsKActOL1nJf

File: source/client/src/clientMain.c [hunk 3]
```diff
@@ -2166,9 +2166,37 @@ int taos_stmt2_bind_param(TAOS_STMT2 *stmt, TAOS_STMT2_BINDV *bindv, int32_t col
...
+  }
+
   int32_t code = 0;
   for (int i = 0; i < bindv->count; ++i) {
     if (bindv->tbnames && bindv->tbnames[i]) {
+      if (pStmt->sql.stbInterlaceMode) {
+        if (tSimpleHashGet(hashTbnames, bindv->tbnames[i], strlen(bindv->tbnames[i])) != NULL) {
+          tSimpleHashCleanup(hashTbnames);
+          tscError("stmt2 bind failed: duplicate table name %s is not allowed in interlace mode.", bindv->tbnames[i]);
+          terrno = TSDB_CODE_PAR_TBNAME_DUPLICATED;
```
Comment (Pengrongkun) [inline source/client/src/clientMain.c]: 同上

File: source/common/src/tdataformat.c [hunk 4]
```diff
@@ -3317,6 +3339,24 @@ int32_t tRowBuildFromBind2(SBindInfo2 *infos, int32_t numOfInfos, bool infoSorte
       code = terrno;
       goto _exit;
     }
+
+    if (orderedDup) {
+      tRowGetKey(row, &rowKey);
+      if (iRow == 0) {
+        // init to ordered by default
```
Comment (Pengrongkun) [inline source/common/src/tdataformat.c]: jeff说过代码不要写注释，尽量通过命名规范来替代注释。建议将二维数组换成ordered和duplicateTs两个bool变量，容易阅读维护

**Review by Pengrongkun — COMMENTED** [review]
修改下review，再跑一下CI

File: source/client/src/clientMain.c (hunk 1 above)
Comment (haolinw) [reply source/client/src/clientMain.c:2229]: Done.

File: source/client/src/clientMain.c (hunk 2 above)
Comment (haolinw) [reply source/client/src/clientMain.c]: Done.

File: source/client/src/clientMain.c (hunk 3 above)
Comment (haolinw) [reply source/client/src/clientMain.c]: Done.

File: source/common/src/tdataformat.c (hunk 4 above)
Comment (haolinw) [reply source/common/src/tdataformat.c]: Done.

**Review by Pengrongkun — APPROVED** [review]

**Review by guanshengliang — APPROVED** [review]

**Pull request merged**